components/percona_telemetry/data_provider.cc

```diff
@@ -0,0 +1,511 @@
+
+  // starting from this point, if the above succeeded we need to close mysql_h.
+  std::shared_ptr<bool> mysql_h_close_guard(
+      new bool(), [&srv = command_factory_service_, &mysql_h,
+                   do_close = !sstatus](bool *p) {
+        delete p;
+        if (do_close) srv.close(mysql_h);
+      });
+
+  // if any of the above failed, just exit
+  if (sstatus) {
+    goto err;
+  }
+
+  if (command_query_service_.query(mysql_h, query.data(), query.length())) {
+    if (err_no) {
+      command_error_info_service_.sql_errno(mysql_h, err_no);
+      if (command_error_info_service_.sql_errno(mysql_h, err_no)) {
+        logger_.warning("Failed to get the last query error");
+      }
+    }
+    goto err;
+  }
+
+  command_query_result_service_.store_result(mysql_h, &mysql_res);
+  if (mysql_res) {
+    std::shared_ptr<bool> query_result_free_guard(
+        new bool(),
+        [&srv = command_query_result_service_, &mysql_res](bool *p) {
+          delete p;
+          srv.free_result(mysql_res);
+        });
+
+    if (command_query_service_.affected_rows(mysql_h, &row_count)) {
+      goto err;
+    }
+    if (command_field_info_service_.num_fields(mysql_res, &num_column)) {
+      goto err;
+    }
+
+    for (uint64_t i = 0; i < row_count; i++) {
+      if (command_query_result_service_.fetch_row(mysql_res, &row)) {
+        goto err;
+      }
+      ulong *length = nullptr;
+      command_query_result_service_.fetch_lengths(mysql_res, &length);
+
+      Row new_row;
+
+      for (unsigned int j = 0; j < num_column; j++) {
+        new_row.emplace_back(std::string(row[j]));
+      }
+      result->push_back(std::move(new_row));
+    }
+  }
+
+  res = false;
+err:
+  if (res && !suppress_query_error_log) {
+    logger_.info("do_query() failed. query: %s", query.c_str());
+  }
+  return res;
+}
+
+const std::string &DataProvider::get_database_instance_id() {
+  if (!database_instance_id_cache_.length()) {
+    QueryResult result;
+    if (do_query("SELECT @@server_uuid", &result)) {
+      static std::string empty;
+      return empty;
+    }
+    database_instance_id_cache_ = result[0][0];
+  }
+  return database_instance_id_cache_;
+}
+
+bool DataProvider::collect_db_instance_id_info(rapidjson::Document *document) {
+  const std::string &id = get_database_instance_id();
+
+  /* Id can be empty if:
+     1. There is too low grace interval and the server didn't start yet,
+        so the SQL query failed. It will recover next time.
+     2. Some other reason that caused selecting server_id to fail. */
+  if (id.length() == 0) {
+    logger_.warning(
+        "Collecting db_instance_id failed. It may be caused by server still "
+        "initializing.");
+    return true;
+  }
+  rapidjson::Document::AllocatorType &allocator = document->GetAllocator();
+  rapidjson::Value instance_id;
+  instance_id.SetString(id.c_str(), allocator);
+  document->AddMember(rapidjson::StringRef("db_instance_id"), instance_id,
```

Comment:
Here and everywhere, ideally the JSON key names should be defined as constants in a single place.

Reply:
done